rust_src/crates/webrender/src/frame.rs

```diff
@@ -1,126 +1,41 @@
-        .outputs
-        .insert(window_id, frame.wr_output());
-
-    frame
-}
-
-pub fn frame_edges(frame: LispObject, type_: LispObject) -> LispObject {
-    let frame = window_frame_live_or_selected(frame);
-
-    let output = frame.wr_output();
-
-    let window = output.get_window();
-
-    let (left, top, right, bottom) = match type_ {
-        Qouter_edges => {
-            let pos = window
-                .outer_position()
-                .unwrap_or_else(|_| PhysicalPosition::<i32>::new(0, 0));
-
-            let size = window.outer_size();
-
-            let left = pos.x;
-            let top = pos.y;
-            let right = left + size.width as i32;
-            let bottom = top + size.height as i32;
-
-            (left, top, right, bottom)
-        }
-        Qinner_edges => {
-            let pos = window
-                .inner_position()
-                .unwrap_or_else(|_| PhysicalPosition::<i32>::new(0, 0));
-            let size = window.inner_size();
-            let internal_border_width = frame.internal_border_width();
-
-            // webrender window has no interanl menu_bar, tab_bar and tool_bar
-            let left = pos.x + internal_border_width;
-            let top = pos.x + internal_border_width;
-            let right = (left + size.width as i32) - internal_border_width;
-            let bottom = (top + size.height as i32) - internal_border_width;
-
-            (left, top, right, bottom)
-        }
-        // native edges
-        _ => {
-            let pos = window
-                .inner_position()
-                .unwrap_or_else(|_| PhysicalPosition::<i32>::new(0, 0));
-            let size = window.inner_size();
-
-            let left = pos.x;
-            let top = pos.y;
-            let right = left + size.width as i32;
-            let bottom = top + size.height as i32;
-
-            (left, top, right, bottom)
-        }
-    };
-    unsafe { list4i(left as i64, top as i64, right as i64, bottom as i64) }
-}
+use crate::gl::context::GLContextTrait;
+use crate::output::CanvasRef;
+use crate::window_system::frame::FrameID;
+use crate::window_system::output::OutputRef;
+use emacs::frame::LispFrameRef;
+use raw_window_handle::RawDisplayHandle;
+use raw_window_handle::RawWindowHandle;
+use webrender::api::ColorF;
+use webrender::{self, api::units::*};
+
+use super::display_info::DisplayInfoRef;
 
 pub trait LispFrameExt {
-    fn wr_output(&self) -> OutputRef;
-    fn wr_display_info(&self) -> DisplayInfoRef;
+    fn output(&self) -> OutputRef;
+    fn canvas(&self) -> CanvasRef;
+    fn set_cursor_color(&self, color: ColorF);
+    fn cursor_color(&self) -> ColorF;
+    fn cursor_foreground_color(&self) -> ColorF;
+    fn set_background_color(&self, color: ColorF);
+    fn display_info(&self) -> DisplayInfoRef;
+    fn window_handle(&self) -> Option<RawWindowHandle>;
+    fn display_handle(&self) -> Option<RawDisplayHandle>;
+    fn size(&self) -> DeviceIntSize;
+    fn uuid(&self) -> FrameID;
```

Comment:
I think "id" or "unique_id" is better here if FrameId is not a UUID(https://en.wikipedia.org/wiki/Universally_unique_identifier)